aten/src/ATen/native/cuda/ForeachTensorAddScalar.cu

```diff
@@ -0,0 +1,90 @@
+        int chunk_size,
+        TensorListMetadata<2>& tl,
+        x_t scalar) {
+            int tensor_loc = tl.block_to_tensor[blockIdx.x];
+            int chunk_idx = tl.block_to_chunk[blockIdx.x];
+            int n = tl.sizes[tensor_loc];
+
+            x_t* x = (x_t*)tl.addresses[0][tensor_loc];
+            x += chunk_idx * chunk_size;
+
+            out_t* out = (out_t*)tl.addresses[1][tensor_loc];
+            out += chunk_idx * chunk_size;
+
+            n -= chunk_idx * chunk_size;
+
+            x_t r_x[kILP];
+            out_t r_out[kILP];
+
+            // to make things simple, we put aligned case in a different code path
+            if(n % kILP == 0 && chunk_size % kILP == 0 && is_aligned(x) && is_aligned(out)) {
+                for(int i_start = threadIdx.x; i_start * kILP < n && i_start * kILP < chunk_size; i_start += blockDim.x) {
+                    // load
+                    load_store(r_x, x, 0 , i_start);
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<x_t>(r_x[ii]) + scalar;
+                    }
+                    // store
+                    load_store(out, r_out, i_start, 0);
+                }
+            }
+            else {
+                // Non-divergent exit condition for __syncthreads, not necessary here
+                for(int i_start = 0; i_start < n && i_start < chunk_size; i_start += blockDim.x * kILP) {
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_x[ii] = 0;
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size) {
+                            r_x[ii] = x[i];
+                        }
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<x_t>(r_x[ii]) + scalar;
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size)
+                            out[i] = r_out[ii];
+                    }
+                }
+            }
+        }
+};
+
+} // namespace
+
+std::vector<Tensor> foreach_tensor_add_scalar_kernel_cuda(TensorList tensors, Scalar scalar) {
+    std::vector<std::vector<at::Tensor>> tensor_lists; 
+    std::vector<at::Tensor> vec_res;
+    for (int i = 0; i < tensors.size(); i++) {
+        vec_res.emplace_back(at::empty_like(tensors[i]));
+    }
+
+    tensor_lists.emplace_back(std::move(tensors.vec()));
+    tensor_lists.emplace_back(std::move(vec_res));
+
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kBool, kBFloat16, kHalf, tensors[0].scalar_type(), "foreach_tensor_add_scalar_kernel_cuda", [&]() {
+        multi_tensor_apply<2>(tensor_lists, AddScalarFunctor<scalar_t, scalar_t>(), scalar.to<scalar_t>());
```

Comment:
This is still not doing correct type promotion behavior: if tensors in TensorList are int, and scalar is float, then doing it in pytorch (or via cpu loop implementation) will result in float tensors, and here the result would still be int. You should fix it and add a test for it.

Reply:
For now I suggest instead of doing correct type promotion just check if scalar is integral type if tensor is integral type - that should fix this error.

Reply:
done